### ISSUE TITLE
Add support for the redcap_custom_verify_username hook and/or hook return values

### ISSUE THREAD

**Issue by mmcev106** [vanderbilt/redcap-external-modules#89]
See this discussion:

https://community.projectredcap.org/questions/36986/external-modules-version-of-redcap-custom-verify-u.html?childToView=37095

**Comment by 123andy** [comment]
I just beat my head on the wall (and total forgot about this thread/issue).  @lsgs 

I couldn't get it working and starting playing around trying to figure out why the 'return' value wasn't getting out of the EM.  I made a few tweaks to the EM code so that it made its way out but found that this still did not solve the problem.  

There was another change made to REDCap (probably some time ago) where the way that this hook is 'handled' - it no longer processes the 'status' and 'message' as is specified in the docs.   Basically, it ignores the return value.

So, here we are:
- I did a little hacking to the EM code to store return values from callling hooks, then filter them and based on the responses throw an email if more than 1 module tries to return a value for a hook like this.  But, REDCap doesn't honor the responses any more so what does this solve?

- Alternately, you can 'render' the message inside the hook and then, if you want to cancel (e.g. status = false, you can call an 'exit()' but this throws an EM error email...

So, we need to fix one or both of these things.  The quick and dirty fix is to render the code inside the hook and then call 'exit' for status-> false.

@taylorr4  - why was the behavior of this changed to ignore the return value of the hook?  There is even an example in the docs that suggest that you might be using this for Vanderbilt LDAP - is that not true?

-Andy

**Comment by 123andy** [comment]
I decided to bail on the return values from the hook and instead just added a method to gracefully exit in the event of a 'failed' lookup, e.g. `[ 'status'=>false ]` which seems to be working.

I did have work-around to return the values but it appears that the core REDCap code no longer handles them so it is irrelevant.  

For now, I'd like to get the pull request accepted.  Going forward, though, I think we should have a system where a hook could return a structured object that would drive REDCap, so making it possible to return values form an EM hook is important. 

Since hooks are evaluated across all modules, it is possible that many modules could have return values that would have to be handled by the core redcap code.  How to do this?

1) We could throw a warning if more than one module tries to control the 'result' object of a hook.  
2) We could return an array of return values, where each hook that has a return value is a member of the resulting array.  The problem here is that I dont know how REDCap core code would handle this.  In most cases, a return value is expected to be a single object.

So, I say that we capture the result.  Something like:

In startHook:
```
		$result = null;
		foreach ($hookNames as $thisHook) {
			if(method_exists($instance, $thisHook)){
				self::setActiveModulePrefix($prefix);
				try{
					$result = call_user_func_array(array($instance,$thisHook), $arguments);
				}
				catch(Exception $e){
					$message = "The '" . $prefix . "' module threw the following exception when calling the hook method '".$thisHook."':\n\n" . $e;
					error_log($message);
					ExternalModules::sendAdminEmail("REDCap External Module Hook Exception - $prefix", $message, $prefix);
				}
				self::setActiveModulePrefix(null);
				continue; // No need to check for the alternate hook name.
			}
		}
		$instance->setRecordId(null);
		return $result;
```

Then, in callHook we have to aggregate all of those results:
```
			$resultsByPrefix = [];  // A new array to hold the results for each module
			foreach($versionsByPrefix as $prefix=>$version) {
			    $resultsByPrefix[] = self::startHook($prefix, $version, $arguments);
			}

			$callDelayedHooks = function($lastRun) use ($arguments){
				$prevDelayed = self::$delayed[self::$hookBeingExecuted];
				self::$delayed[self::$hookBeingExecuted] = array();
				self::$delayedLastRun = $lastRun;
				foreach ($prevDelayed as $prefix=>$version) {
					$resultsByPrefix[] = self::startHook($prefix, $version, $arguments);
				}
			};
```

Now we have an array of $resultsByPrefix for the hook.  Since most will be empty, we can remove those:

```
           // Clean up empty results
            $resultsByPrefix = array_filter($resultsByPrefix);
```

Then we just need to take the first value and throw a warning if there are more than one value:
```
            // Throw and email if someone has two modules using this hook
            if (count($resultsByPrefix) > 1) {
                // We have more than one EM returning results for a hook.
                $message = "You have more than one EM (" . implode(",", array_keys($resultsByPrefix)) .
                    ") that are using the $name hook to return a result.  Only the first response will be returned.";
                    ExternalModules::sendAdminEmail("REDCap External Module Exception", $message, $prefix);
            }

            if (count($resultsByPrefix > 0)) {
                // Return the first result
                reset($resultsByPrefix);
                return current($resultsByPrefix);
            }
```

The part where it gets a little murkier, is the fact that each 'hook' is really a combination of traditional hooks AND external module hooks.

So, in the HOOKS class `call` method, we really have to do the same thing all over again.  We have to keep track of results from the various ways we call a hook and then reduce it down to a single possible return value...   (code not shown).

I think this would work..  Thoughts?

**Comment by taylorr4** [comment]
@123andy none of that hook's behavior was ever "changed" that i know of. To my knowledge, since the redcap_custom_verify_username hook was always different than the other hooks, it simply never worked with External Modules at all. I haven't had a chance to dig through all your code suggestions, but i'm sure anything at this point would be an improvement for that hook.

**Comment by 123andy** [comment]
Ahh.  Sorry, @taylorr4.  I found the exit code in the Hooks.php file...  I was looking in the wrong place assuming that this return value made it back into the calling REDCap file (e.g. assign_user.php)...  

With this 'new realization' I think the 'allowExit' solution is probably the easiest for now.  It also gives the developer more flexibility with how REDCap handles this particular hook...  

This hook is an oddity in that it is in the only one that results in calling the 'actual' method in the Hooks.php class...  I didn't realize that earlier.

So, what hooks do NOT support is something like this:
```
if (!Authentication::isTableUser($user)) {
  $hookResult = Hooks::call('redcap_custom_verify_username', array($user));
  // Based on $hookResult, modify behavior of source code...
}
```
Do we need this?  Maybe not -- made it this far.  Heck, we could just drop the entire '$result' concept from hooks and tell people for the redcap_custom_verify_username hook to just do something analogous to what you are doing in the hook function...

Hopefully, my EM can help replace the need for this custom hook anyhow...

**Comment by mmcev106** [comment]
Moved from PR #164:

Per past discussions, I'm not sure we can or should ever support exiting in the middle of a hook, or any other action that allows one module to change the behavior of another module without some kind of consent from the latter module.

In a few hours, I believe I could made the required changes in REDCap and the module framework to allow return codes to work with the **redcap_custom_verify_username** hook. I think the _delayModuleExecution()_ method might even be used to allow multiple modules to run **redcap_custom_verify_username** on the same project. I believe this solution will be easier to support long term. If this is holding you back, would you like me to see if I can get the OK to work on this?

**Comment by mmcev106** [comment]
If we ever encounter an actual use case where it makes sense to enable two modules implementing **redcap_custom_verify_username**, I think it should be up to those modules to determine the correct behavior (the framework likely shouldn't dictate whether the **status** values from each are **AND**ed or **OR**ed).  Using **delayModuleExecution()** has worked very well in previous undefined run order scenarios.  Perhaps we should have a **getLastReturnValue()** method or something that can be used to retrieve the value returned from the previous module's hook from the current module's hook.  That would also even allow custom behavior based on other return values (like the **message** for **redcap_custom_verify_username**).

**Comment by 123andy** [comment]
I think the avoidance of exit is great in theory but in practice it present many challenges.  As an example, imagine you have two hooks that redirect after a form is saved to an EM plugin page.  Since you can't just render the page and exit, you have to do a client-based redirection method of pushing javascript.  We've had to do this a few times.  If you had two conflicting EMs trying to do the same thing it would be really hard to determine what was at fault and why one was happening and not the other.

I think allowing 'exit' but making it explicit (as this does) is another tool that can be used as needed understanding its limits.  Yes, it is a dead-end, but sometimes the code you want to push in an EM is a dead-end that isn't compatible with other modules on the same hook.  Does this mean the functionality is invalid?  Perhaps we could say it has to be a delayed execution hook as well?

I do think we should also address the return issue - but it doesn't have to be an either-or.  It probably makes sense to array up return values from all the modules and let the hook decide how it handles it.  There probably are cases where the hook could handle multiple inputs and others where it would just take the last if more than one was presented...

Right now, I want to get this hook out as we are running into issues and I don't want to be patching the EM code :-)

**Comment by mmcev106** [comment]
It is currently possible for two modules to coordinate and decide which should trigger redirection by storing/checking field values and using **delayModuleExecution()**.

Supporting **exit()** would make it easy to unexpectedly introduce scenarios where one module could prevent others from running, and the behavior would be undefined/inconsistent as to whether other modules would run prior to the **exit()** call.  As the primary gatekeeper to make sure we can sustainably support our community, I cannot in good conscience allow this based on all the information and experience I currently have.  I'm happy to discuss this further, but would likely need new information for my perspective to change.

I am going to try to prioritize adding return value support for you.  I will report back within the next couple of days.

**Comment by 123andy** [comment]
Hi Mark - I just added a pull request for handling return variables for hooks:  #166   Please take a look and let me know what you think.

We can save the discussion on module interop for a beer at REDCap Con :-)

Thx!
-A

**Comment by mmcev106** [comment]
PR #166 has been merged, as well as [PR #92 on REDCap Core](https://github.com/vanderbilt/REDCap/pull/92) per your comments on #166.  These changes should make it into the next release after 3/11.

**Comment by mmcev106** [comment]
I'm looking forward to hearing more or your module intertop thoughts!  Closing...
